### PR TITLE
PyTorch & MXNet-Python Inference Script Demo

### PR DESCRIPTION
**Pull request by HeliWang**
The PR contains minimal inference example of the following items:
   - ONNX model
   - script to run Kim model in PyTorch
   - script to run ONNX Kim model in MXNet (Python)

  Will add script to run ONNX Kim model in MXNet (Scala) in later PR after this PR is approved.

### CODE REVIEW

**Comment by lintool** [comment]
@Impavidity can you do CR please?

**Comment by lintool** [comment]
@HeliWang `word2vec.sst-1` - Are these just the word vectors in SST?
If so, we should support access to the all word vectors so that we can support a real demo (i.e., user can type in arbitrary text)?
We don't actually want to check all the word vectors in the repo, so a script that wgets the data from a known location?

File: embedding.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+def embedding(sentences, length, single_sentence_length):
+    """
+    sentences is a 2d list, list<list<word>> for inference
+    if length > len(sentences), padding 0 to extend the matrix
+    """
+    matrix = np.zeros((length, 1, single_sentence_length, 300))
+
+    for i in range(len(sentences)):
+        words = sentences[i]
+        sentence_matrix = sentence_to_embedding(words, single_sentence_length, matrix, i)
```
Comment (Impavidity) [inline embedding.py]: The return value `sentence_matrix` is never used? Could delete the variable.

File: kim_cnn/SST1.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+class SST1Dataset(data.TabularDataset):
+    dirname = 'data'
+    @classmethod
+    def splits(cls, text_field, label_field,
+               train='phrases.train.tsv', validation='dev.tsv', test='test.tsv'):
+        prefix_name = 'stsa.fine.'
+        path = './kim_cnn/data'
+        return super(SST1Dataset, cls).splits(
+            path, train=prefix_name + train, validation=prefix_name + validation, test=prefix_name + test,
+            format='TSV', fields=[('label', label_field), ('text', text_field)]
```
Comment (Impavidity) [inline kim_cnn/SST1.py]: This script could not support current version of torchtext, could you refer the current code in Kim CNN and adapt this script ?
- format='TSV' should be format='tsv'

Do you have the script to build the docker? Could we also put that into the repo?

File: kim_cnn/train.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+            # switch model into evalutaion mode
+            model.eval(); dev_iter.init_epoch()
+            n_dev_correct = 0
+            dev_losses = []
+            for dev_batch_idx, dev_batch in enumerate(dev_iter):
+                model_input = embedding.embed(dev_batch)
+                scores = model(model_input)
+                n_dev_correct += (torch.max(scores, 1)[1].view(dev_batch.label.size()).data == dev_batch.label.data).sum()
+                dev_loss = criterion(scores, dev_batch.label)
+                dev_losses.append(dev_loss.data[0])
```
Comment (Impavidity) [inline kim_cnn/train.py:153]: adapt the torch code to PyTorch 0.4
- dev_loss.data[0] should be modified to dev_loss.item()
refer to https://github.com/castorini/Castor/blob/master/common/trainers/sst_trainer.py#L46 for more info

**Comment by Impavidity** [comment]
- Try to use PEP8 to reformat the code
- Let's upgrade to software version

File: kim_cnn/SST1.py (hunk 2 above)
Comment (HeliWang) [reply kim_cnn/SST1.py]: Many thanks for the review! Will reply the feedback later.

For the docker image source code:
https://github.com/HeliWang/pytorch-mxnet-jvm

The version of pytorch/torchtext I used:
  torch 0.3.1
  torchtext 0.2.3

**Comment by tuzhucheng** [comment]
Hi @HeliWang it seems you are using an older version of the Kim CNN implementation in Castor?

**Comment by Impavidity** [comment]
@tuzhucheng This version is different from the one in the repo. Are we going to adapt this to the latest version? @HeliWang Can you describe what's the issue with the pytorch 0.4 and ONNX here ?

**Review by Impavidity — APPROVED** [review]

**Review by tuzhucheng — APPROVED** [review]

**Pull request merged**